### ISSUE TITLE
Apparent Non-Standard Behavior For MediaTek Hardware RZ608

### ISSUE THREAD

**Issue by thesunexpress** [pgj/freebsd-wifibox#24]
Having an issue getting the current wifibox image detecting a NIC that has been, as far as I can tell, included since kernel 5.12.
wifibox's dmesg & lspci do not even show this NIC being present... Yet when testing with Arch Linux (kernels: 5.12 / 5.15 / 5.17) this NIC is immediately detected & active. 

I've noted that parts of the common toolset in current wifibox are also missing. I figured I'd try and manually update the image, with an apk update etc, only to find out that these tools are all gone! So I guess manually adding even additional firmware via apk add is totally out of the question now. 
Any particular reason why the apk tool was completely removed? 

Hardware in question:   MEDIATEK Corp. RZ608 Wi-Fi 6E 80MHz 
Arch Linux detects this device as  "mt7921e", which I assumed would be covered by the wifibox-alpine port with FW_MEDIATEK. But no luck so far.

There appears to be, hopefully unrelated, an issue when building wifibox from source. Compilation seems to proceed correctly, but I do note the following being thrown at the console:
" linux: jid 0 pid 1932 (apk): unsupported umount2 flags 0xa "
The fact that apk is not present in the image, yet seems to make some form of appearance here, is confusing. Also notable that linprocfs.ko gets KLDloaded when building the net/wifibox-alpine port... not entirely sure why this occurs. Can't find anything in the manual(s) indicating why this might be. Neither does it mention anything about the need to KLDload linux.ko or linux64.ko which are obviously required.

Any hints as to how I might get the correct firmware and/or kernels bits injected into the wifibox image?

**Comment by pgj** [comment]
There was a big (breaking) change about how the guest image is built.  Check out the [freebsd-wifibox-alpine](https://github.com/pgj/freebsd-wifibox-alpine) repository for the details, as well as the corresponding [`net/wifibox-alpine` port](https://github.com/pgj/freebsd-wifibox-port) for the details, but I will briefly summarize the main changes here.

For the smaller size, better security and automation, the guest image is now always built from scratch with SquashFS which is a read-only compressed file system therefore it does not allow modifications after it has been constructed.  The way to add more packages is to change the `net/wifibox-alpine` port's [`Makefile`](https://github.com/pgj/freebsd-wifibox-port/blob/579a20a3a59f9c9171bb54d581c9e8d28918b860/net/wifibox-alpine/Makefile#L100) where the deployed APKs are listed.  APKs are only used to put together the contents of the images, but [some of them are stripped down and optimized for Wifibox](https://github.com/pgj/freebsd-wifibox-alpine/tree/main/aports) therefore one cannot just update from the Alpine package repository.

The `apk` tool is used for building the `net/wifibox-alpine` port, but, to avoid the need for `chroot(8)` so that the port could be built as a user, the `apk` binary is made to work with the help of Linuxulator, that is why it needs `linux64` and co.  I wrote about this in the [`README`](https://github.com/pgj/freebsd-wifibox-port/blob/579a20a3a59f9c9171bb54d581c9e8d28918b860/README.md) but that is indeed not visible when one tries to build the port from the Ports Collection directly.  The error about `unsupported umount2 flags` can be ignored, `apk` will still work as much as it is required.

The kernel bits for the driver are coming from the custom-built `linux-lts` package, along [its configuration](https://github.com/pgj/freebsd-wifibox-alpine/blob/8786c997a36ec769c22e081358f341072f59bffa/aports/linux-lts/config-lts.x86_64) and the firmware bits added in the port's [Makefile](https://github.com/pgj/freebsd-wifibox-port/blob/579a20a3a59f9c9171bb54d581c9e8d28918b860/net/wifibox-alpine/Makefile#L253) from the `linux-firmware` snapshot or the [MediaTek firmware package](https://github.com/pgj/freebsd-wifibox-port/blob/579a20a3a59f9c9171bb54d581c9e8d28918b860/net/wifibox-alpine/Makefile#L222).  Those need to be modified if one wants to fix that part.  I believe the kernel driver is there, it is firmware that is missing.

I do not have a MediaTek device that why I could not test if it is working therefore there might be things that I forgot to add to the image.  But with your help this can be fixed.  Could you share me some log snippets where the hardware discovery is logged, e.g. `dmesg`?

**Comment by thesunexpress** [comment]
Excellent work! You've done some serious configuring of that LTS image... impressive.
Looking through the config & Makefile, I don't see any reason why the MediaTek bits get omitted. Must be something more fundamental with the base image.
Do you need the `dmesg` output from the (not-working) Alpine image as well? Or the the (working) Arch Linux as well? 

Here's the FreeBSD output:
`pciconf`
ppt0@pci0:5:0:0:	class=0x028000 rev=0x00 hdr=0x00 vendor=0x14c3 device=0x0608 subvendor=0x14c3 subdevice=0x0608
    vendor     = 'MEDIATEK Corp.'
    device     = 'RZ608 Wi-Fi 6E 80MHz'
    class      = network

`dmesg`
<snip>
pcib5: <ACPI PCI-PCI bridge> at device 4.0 on pci3
pci5: <ACPI PCI bus> on pcib5
pci5: <network> at device 0.0 (no driver attached)
pcib6: <ACPI PCI-PCI bridge> at device 8.0 on pci3
pci6: <ACPI PCI bus> on pcib6
pci6: <unknown> at device 0.0 (no driver attached)
<snip>
ugen0.2: <ITE Tech. Inc. ITE Device> at usbus0
kbd2 at ukbd0
Root mount waiting for: usbus0
ugen0.3: <MediaTek Inc. WirelessDevice> at usbus0   <-----?!?!
Setting hostuuid: 035e02d8-hello-how-are-you12345678.
<snip>

Interestingly, it is the Bluetooth part of the hardware that gets recognized as USB for some reason by the FreeBSD kernel.
The Arch Linux install, I have Bluetooth completely removed & the NIC part of the device still works perfectly without issue; no indication at all why Linux doesn't consider it a USB device either --- it never showed up as a USB there, even before Bluetooth amputation.

Let me know if you need any extra information.

**Comment by pgj** [comment]
Thanks for the FreeBSD output.  Bluetooth is not (yet) in scope hence I will ignore those details for the moment.

What I would like to really see is the `dmesg` output of the Alpine guest.  Whether the driver's kernel module is probed for loading, whether gives an message, complains about something, etc.  From the `wifibox console`, it might be interesting to which module got loaded eventually (if it could), i.e. the output of the `lsmod` command.  For the reference, it might be useful to include the same set of output that are extracted from the native Arch Linux.

As far as I understand the situation, this card should be handled by the `mt7921e` driver, which is available in the image.  According to `modinfo`, this driver needs to the following firmware files:

~~~
mediatek/WIFI_MT7961_patch_mcu_1_2_hdr.bin
mediatek/WIFI_RAM_CODE_MT7961_1.bin
~~~

These are provided by the `linux-firmware` snapshot and are added to the guest image when the `FW_MEDIATEK` option is set.  Another potential source of problem could be the bhyve PCI pass-through implementation which might fail to work in this case.  What FreeBSD version are you using?

**Comment by thesunexpress** [comment]
Forgive the delay, real life things got in the way...

Here is the `dmesg` & such of the `wifibox` in question:

https://pastebin.com/vyvtwpku

**Comment by thesunexpress** [comment]
And here's the FreeBSD box that hosts the `wifibox` install:

https://pastebin.com/nMZ5qYJ0

As can be noted, `wifibox` indicates `00:06.0 Class 0280: 14c3:0608` with no driver attached.
Which is precisely the device found at `ppt0@pci0:5:0:0:	class=0x028000 rev=0x00 hdr=0x00 vendor=0x14c3 device=0x0608 subvendor=0x14c3 subdevice=0x0608
`
Having built `wifibox` with both the default drivers included and/or customizing which drivers are built & injected, results in an image without driver support for this device. I'll hop over to the Arch Linux install & pull info from there in a moment.

**Comment by pgj** [comment]
Thanks!  Looks like no Linux driver is loaded at all.  Not even tried for loading.  I have just noticed that this is an AMD CPU, where PCI pass-through has to be enabled explicitly.  Have you done that?  That is, issue the following command:

```console
# sysctl hw.vmm.amdvi.enable=1
```

before loading the `vmm(4)` kernel module.  You can also try to load the Linux driver on the Wifibox console:

```console
# modprobe mt7921e
```

**Comment by thesunexpress** [comment]
Hah, pastebin freaks out about the following content, claiming there's something offensive about it.
"Pastebin’s SMART filters have detected potentially offensive or questionable content in your paste." 
Not so 'smart' after all ¯\_(ツ)_/¯  who knows. 
Here goes:

modinfo mt7921e                                                                                                                                                               
filename:       /lib/modules/5.18.3-arch1-1/kernel/drivers/net/wireless/mediatek/mt76/mt7921/mt7921e.ko.zst
license:        Dual BSD/GPL
author:         Lorenzo Bianconi <lorenzo@kernel.org>
author:         Sean Wang <sean.wang@mediatek.com>
firmware:       mediatek/WIFI_MT7922_patch_mcu_1_1_hdr.bin
firmware:       mediatek/WIFI_RAM_CODE_MT7922_1.bin
firmware:       mediatek/WIFI_MT7961_patch_mcu_1_2_hdr.bin
firmware:       mediatek/WIFI_RAM_CODE_MT7961_1.bin
srcversion:     F7F2D6D96C96FE320DF7007
alias:          pci:v000014C3d00000616sv*sd*bc*sc*i*
alias:          pci:v000014C3d00000608sv*sd*bc*sc*i*
alias:          pci:v000014C3d00007922sv*sd*bc*sc*i*
alias:          pci:v000014C3d00007961sv*sd*bc*sc*i*
depends:        mt76,mt7921-common,mt76-connac-lib
retpoline:      Y
intree:         Y
name:           mt7921e
vermagic:       5.18.3-arch1-1 SMP preempt mod_unload 
sig_id:         PKCS#7
signer:         Build time autogenerated kernel key
sig_key:        1E:AB:6F:E7:59:FF:1F:DD:02:2A:E1:2F:2D:56:48:0B:BF:D0:E5:80
sig_hashalgo:   sha512
signature:      30:65:02:30:30:1D:30:E5:32:4C:0B:5C:A2:63:6F:F5:07:3A:39:A1:
		46:D6:85:66:20:F1:F8:9A:1B:9C:D8:51:32:F7:87:65:14:76:D9:7C:
		D6:BC:A5:4D:8C:51:43:04:77:2D:79:BB:02:31:00:F2:6C:84:54:3F:
		EC:03:0E:C8:B1:E6:06:F0:63:38:CC:75:DF:A6:DA:8D:A1:04:D9:86:
		19:A6:24:8F:D2:4E:86:D4:57:5A:12:D0:04:20:79:6D:B1:8B:9E:CD:
		29:BB:4E
parm:           disable_aspm:disable PCI ASPM support (bool)


enp4s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.LOL!  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 nyet  prefixlen 64  scopeid 0x20<link>
        inet6 re-nyet  prefixlen 128  scopeid 0x0<global>
        ether jaguaryu  txqueuelen 1000  (Ethernet)
        RX packets 51331  bytes 72987923 (69.6 MiB)
        RX errors 0  dropped 1  overruns 0  frame 0
        TX packets 8183  bytes 844673 (824.8 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device memory 0xfc400000-fc4fffff  

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlan0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.brb  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 boom-de-yadda  prefixlen 128  scopeid 0x0<global>
        inet6 boom-de-yadda  prefixlen 64  scopeid 0x20<link>
        ether ilovethewholeworld  txqueuelen 1000  (Ethernet)
        RX packets 1528  bytes 159999 (156.2 KiB)
        RX errors 0  dropped 278  overruns 0  frame 0
        TX packets 109  bytes 7298 (7.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

**Comment by pgj** [comment]
But is the `mt7921e` driver that is loaded on Arch Linux?  What does `lsmod` say there?

**Comment by thesunexpress** [comment]
>I have just noticed that this is an AMD CPU, where PCI pass-through has to be enabled explicitly. Have you done that? That is, issue the following command: 
> # sysctl hw.vmm.amdvi.enable=1
 > before loading the `vmm(4)` kernel module. 

Remember I'm the one that told you about this some time ago?  😉

**Comment by pgj** [comment]
Yeah, I had some vague memories about it, but I still wanted to make it clear that it is configured for you in this specific case.

**Comment by thesunexpress** [comment]
> But is the `mt7921e` driver that is loaded on Arch Linux? What does `lsmod` say there?

`lsmod` from the ArchLinux box:

https://pastebin.com/9YTdKwVi

**Comment by pgj** [comment]
Great thanks!  Now the output of `modprobe mt7921e` on Wifibox would be interesting to see.

**Comment by pgj** [comment]
I tried to load `mt7921e` on my instance without any Mediatek hardware present (or published to the VM) and the following modules got loaded in result:

```console
wifibox:~# lsmod
Module                 Size   Used by    Not tainted
mt7921e                81920  0 
mt76_connac_lib        32768  1 mt7921e
mt76                   61440  2 mt7921e,mt76_connac_lib
```

which is a bit different from what one could see in the Arch Linux case.  The `mt7921_common` module seems to be missing:

```console
mt7921_common          94208  1 mt7921e
```

**Comment by pgj** [comment]
This does not look right:

```console
wifibox:~# modprobe mt7921_common
modprobe: module mt7921_common not found in modules.dep
```

**Comment by pgj** [comment]
However, this difference might be explained by the fact that Wifibox is based on Linux 5.15, while your Arch Linux is 5.18.

**Comment by thesunexpress** [comment]
`wifibox:~# modprobe mt7921e
wifibox:~# lsmod
Module                  Size  Used by    Not tainted
mt7921e                81920  0 
mt76_connac_lib        32768  1 mt7921e
mt76                   61440  2 mt7921e,mt76_connac_lib
wifibox:~# 
`

**Comment by thesunexpress** [comment]
What I can say for certain, is that this driver/mod worked on ArchLinux kernels 5.12 / 5.15 / 5.18 <---this last one updated a few hours ago.

**Comment by pgj** [comment]
After loading the `mt7921e` module with `modprobe`, is not there anything mentioned in the logs?  It would be good to see if the driver is actually tried for loading, whether it can do anything with the hardware, etc.

**Comment by thesunexpress** [comment]
So the fact of the matter here is that it is *not* a matter of the firmware missing, but rather that the device is not being detected at all. I wonder what the heck the cause might be...  I don't think it has anything to do with the `bluetooth` section on the device, as this apparently has no influence on anything --- at least when it comes to Intel NICs. I should note that pulling the Mediatek NIC & swapping in an `Intel iwlwifi` device, an `AX200`/ `AX210` specifically, leads to `wifibox` functioning correctly with zero issues.

**Comment by thesunexpress** [comment]
> After loading the `mt7921e` module with `modprobe`, is not there anything mentioned in the logs? It would be good to see if the driver is actually tried for loading, whether it can do anything with the hardware, etc.

So far no complaints. I'll reboot real quick, launch `wifibox` & `modprobe` etc & let it run for a while to see if anything interesting is spit out....

**Comment by pgj** [comment]
According to the change logs, Linux 5.15.42 & 5.15.46 include some fixes that are related to the `mt7921e` driver.  Therefore it might be worth a try to experiment with a more recent version from the 5.15 line.  I have been planning to update the kernel and the firmware files for a while now, and it is a fine excuse to do so now.

**Comment by thesunexpress** [comment]
> According to the change logs, Linux 5.15.42 & 5.15.46 include some fixes that are related to the `mt7921e` driver. Therefore it might be worth a try to experiment with a more recent version from the 5.15 line. I have been planning to update the kernel and the firmware files for a while now, and it is a fine excuse to do so now.

That would be, indeed, an interesting thing to try... and you have at least 1 test bunny here.
 
I suspect it might be something more intricate to this hardware in particular though. At least from a glass-half-empty point of view. It is not a deal-breaker to me, as I can easily swap back to the Intel device, but it would be nice to achieve more hardware coverage --- Mediatek seems to be a popular manufacturer lately. Not to mention the absolutely epic benefits attained by running  `wifibox`.

**Comment by pgj** [comment]
I agree totally.  I definitely want to make Wifibox work on Meditatek cards and I am very grateful for all your help.

**Comment by pgj** [comment]
I have pushed an update for the `net/wifibox` port to GitHub, see https://github.com/pgj/freebsd-wifibox-port/commit/b1945dc17e9d45e1ab963b6f827210cf1ac3e2b4 -- please try it and let me know if it makes any difference.

**Comment by thesunexpress** [comment]
`wifibox:~# uname -a`
`Linux wifibox 5.15.46-0-lts #1-Alpine Sat, 11 Jun 2022 17:55:53 +0000 x86_64 GNU/Linux`

But no love... `modprobe` 'ed mt7921e manually again, device not detected.

**Comment by pgj** [comment]
I have found [a question on Ask Ubuntu](https://askubuntu.com/questions/1380400/no-wifi-adapter-found-on-ubuntu-20-04-3-lts) that might be related.  According to that, not all PCI ids are known by the driver and they have to be configured manually.  This can be learned via the `modinfo` command, and that is what `mt7921e` tells me on Wifibox:

```console
wifibox:~# modinfo mt7921e | fgrep alias
alias:          pci:v000014C3d00007961sv*sd*bc*sc*i*
```

However, in your logs from Arch Linux, the list of ids is as follows.

```console
alias: pci:v000014C3d00000616svsdbcsci*
alias: pci:v000014C3d00000608svsdbcsci*
alias: pci:v000014C3d00007922svsdbcsci*
alias: pci:v000014C3d00007961svsdbcsci*
```

And the id of your card corresponds to the second in the line, I guess...?  This is missing in case of Wifibox and could be an explanation for the problem.

**Comment by pgj** [comment]
In https://github.com/pgj/freebsd-wifibox-port/commit/342c2b049fa70bcb11165e3092fba838d60bb25d I made the changes to `modprobe` that were suggested in the Ask Ubuntu answer.

**Comment by thesunexpress** [comment]
> wifibox:~# modinfo mt7921e | fgrep alias
> alias:          pci:v000014C3d00007961sv*sd*bc*sc*i*
> ```
> alias: pci:v000014C3d00000616svsdbcsci*
> alias: pci:v000014C3d00000608svsdbcsci*
> ```
> 
> And the id of your card corresponds to the second in the line, I guess...? This is missing in case of Wifibox and could be an explanation for the problem.

I can confirm this 100%, as when I boot with this NIC removed, that PCI ID is totally missing. Besides Arch & FreeBSD confirming the same.

**Comment by thesunexpress** [comment]
> In [pgj/freebsd-wifibox-port@342c2b0](https://github.com/pgj/freebsd-wifibox-port/commit/342c2b049fa70bcb11165e3092fba838d60bb25d) I made the changes to `modprobe` that were suggested in the Ask Ubuntu answer.

I get no love from this iteration either. Not sure at all why, such a conundrum. I'm tempted to try and boot/install/run a full Alpine Linux OS as a test to see if the full OS shows any hints.
The kernel module `mt7921e` gets specifically loaded I see, but does not seem to help.

I like the new features & layout you've added. The hostap bits look very very interesting!

**Comment by pgj** [comment]
[Another thing](https://askubuntu.com/questions/1376871/rz608-mt7921k-wireless-lan-driver-is-not-supported-on-ubuntu-18-04/1378043#1378043) I have found that you could try is to "hint" directly the loaded driver:

```console
# modprobe mt7921e
# echo 14c3 0608 > /sys/bus/pci/drivers/mt7921e/new_id
```

**Comment by thesunexpress** [comment]
Bingo.

That worked. I went ahead and did a full test-install of Alpine Linux to see if it behaved any different. It did not. So there's nothing wrong at all with `wifibox`, rather it is some peculiar about Alpine. I used the 'Extended' disk image for the install & it too did not detect this Mediatek NIC. This is rather odd, considering the ease with which the other Linuxes all happily support it out of the box -- in my particular case since at least kernel 5.12.
The contributors at the AskUbuntu.com site do indicate there is some peculiar hardware behavior they've encountered. Strangely, it appears to be different between operating systems too.

So, manually modprobing the `mt7921e` kernel module is still required & `echo 14c3 0608 > /sys/bus/pci/drivers/mt7921e/new_id`  does do the trick in both the full install of Alpine & `wifibox` -- but is not permanent between reboots, as expected.
Creating:
`$ cat /etc/modprobe.d/rz608.conf`
`alias pci:v000014C3d00000608sv*sd*bc*sc*i* mt7921e`
`$ cat /etc/udev/rules.d/99-rz608.rules`
`SUBSYSTEM=="drivers", DEVPATH=="/bus/pci/drivers/mt7921e", ATTR{new_id}="14c3 0608" `
 Does do the trick for Alpine, but not for `wifibox` for obvious & not so obvious reasons. Schrödinger's cat sends his regards. For whatever reason `mount -o remount,rw /` does not place the filesystem in read-write mode at all, so whatever changes I make are not permanent across boots (hooray for the `cache` directories on host & guest) or can't be written to disk at all due to perma-read-only-mode.

Thus I propose to maybe hard-code these files in the FreeBSD port of `wifibox` & adding an entry to the `man` that Mediatek hardware acts weird. Or requires exorcism.

Bonus:  This NIC is significantly faster as a guest on FreeBSD via `wifibox`. Did a bunch of network speedtests, both local & WAN. Finding that under all circumstances 20-25% greater throughput & max speed is achieved with the FreeBSD/`wifibox` combo, compared to vanilla Alpine Linux / Arch Linux / Debian / OpenSUSE. The cause for why a Linux install under-performs, yet to be found out.

**Comment by pgj** [comment]
Great, thanks for testing!  Yes, that is right, the changes have to be permanently hardwired in the Wifibox guest image.  That is what I tried to achieve in https://github.com/pgj/freebsd-wifibox-port/commit/342c2b049fa70bcb11165e3092fba838d60bb25d .  I think the main difference is that `udev` would also be needed -- you had to install the `eudev` package, right?  Another option is to patch the kernel source directly to insert the missing hardware IDs and remove the need for all these user-space configuration steps (also, for the better security hardening).  This sounds like an appealing notion since those IDs should land in the driver eventually anyway.

**Comment by thesunexpress** [comment]
I did not catch it specifically scrolling by when installing some apks in Alpine. I'll check the logs/apk list when I get back. I suspect, obviously it did, or perhaps the Alpine Extended install ISO contains it already -- not installed separately because it was already included.
I guess a hacky way around it is to have both `modprobe mt7921e`  and  `echo 14c3 0608 > /sys/bus/pci/drivers/mt7921e/new_id` run after boot in the guest. Perhaps `/etc/init.d/networking restart` included as well, testing pending. The reason being no udev shenanigans required.

I did notice that wpa_supplicant.conf is neither read nor stored in the guest if wlan0 is never brought up. Not sure what the routine is here, but when I swapped in the Intel AX200 (a `iwlwifi` device) the `wpa_supplicant` passthrough functions as intended.

**Comment by pgj** [comment]
> I did notice that wpa_supplicant.conf is neither read nor stored in the guest if wlan0 is never brought up. Not sure what the routine is here, but when I swapped in the Intel AX200 (a `iwlwifi` device) the `wpa_supplicant` passthrough functions as intended.

Sorry, I do not think I understand what you were referring to exactly.  Perhaps this is the topic of a different issue?

**Comment by thesunexpress** [comment]
> 
I'll do a bit more testing later with the AX200 NIC to see if the behavior still remains. It was peculiar no less, configuration files remain the same, yet when one particular NIC is present the wpa_supplicant.conf file is ignored. Could be something else however, I need to make sure what version(s) are installed first -- but that's for later.

**Comment by pgj** [comment]
@thesunexpress can we say that the MediaTek card now works without issues (per https://github.com/pgj/freebsd-wifibox-port/commit/1234c712c80d0e73ca2b8f710ed3544ca116f5ba) hence this ticket could be closed?

**Comment by thesunexpress** [comment]
Yes indeed!
I'm putting it under heavy load both on LAN & WAN for the next 48 hours, stress testing to see if I can manage to break anything. I doubt there will be any issue since `wifibox` has literally never failed when I do try to throw everything I can at it. The joys of having a 20Gbit fiber home network ;-)

Might be a good idea to add a note to the `man` of an entry regarding the anomalous behavior we got here from MediaTek... I suspect we'll see it again in future hardware from them. This NIC has been on the market for at least 5 months now, but for whatever reason Alpine Linux team still haven't thought about adding the proper PCI ID quirks to the LTS kernel.

**Comment by thesunexpress** [comment]
Indeed. I merely meant to point out that there is in fact nothing wrong with `wifibox`, but rather Alpine. Having a knowledge base of rare obscure peculiar quirky behavior is very useful, as the AskUbunut.com website has made clear. Of the main Linux, currently popular, families like Debian / Fedora (SUSE) / Arch, all seem to have properly included the various ID's since at least kernel 5.12 -- I can't imagine a scenario why Alpine did not / has not yet included them, despite being at 5.15 already. There is probably a good reason, unrelated to `systemd` / `openrc`. In any case, I am quite impressed with Alpine as a whole. Seems to be a nice, tightly packaged modern OS. At a sub-150MB OS, even more impressive.

As for the platform, please absolutely include it in the list of known & fully functional deployments. For what it is worth, both the AMD B550 & AMD X570 platforms, I can confirm full functionality in 13-STABLE, 13.1-STABLE & 14-CURRENT --- along with the Intel AX200 & MediaTek RZ608 NICs. They are 2 different motherboards: ASUS ROG STRIX B550-I GAMING & GigaByte X570S I AORUS PRO AX. They are part of a 144 CPU core cluster computer running in my home lab. Nerd things, I know... They will be running `wifibox` exclusively for the foreseeable future. The new HostAP avenue you are implementing looks very interesting too; I will start exploring this soon.
You can mark this "issue" report as closed. Excellent work!

**Comment by pgj** [comment]
Okay, great.  Thanks for all the input.  Let me know how it goes with `hostapd` -- it was originally motivated by https://github.com/pgj/freebsd-wifibox/issues/22, there you can continue posting about the experiences.
